azurerm/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go

```diff
@@ -0,0 +1,213 @@
+				ServerKeyType: sql.AzureKeyVault,
+				ServerKeyName: &serverKeyName,
+			}
+		} else {
+			return fmt.Errorf("Key vault key id must be a reference to a key, but got: %s", keyDetails.NestedItemType)
+		}
+	} else {
+		serverKeyName = ""
+
+		// Service managed doesn't require a key name
+		encryptionProtectorProperties = sql.EncryptionProtectorProperties{
+			ServerKeyType: sql.ServiceManaged,
+			ServerKeyName: &serverKeyName,
+		}
+	}
+
+	// Only create  aserver key
+	if serverKey.ServerKeyProperties != nil {
+		// Create a key on the server
+		futureServers, err := serverKeysClient.CreateOrUpdate(ctx, resGroup, serverName, serverKeyName, serverKey)
+		if err != nil {
+			return err
+		}
+
+		if err = futureServers.WaitForCompletionRef(ctx, serverKeysClient.Client); err != nil {
+			return err
+		}
+	}
+
+	encryptionProtectorObject := sql.EncryptionProtector{
+		EncryptionProtectorProperties: &encryptionProtectorProperties,
+	}
+
+	futureEncryptionProtector, err := encryptionProtectorClient.CreateOrUpdate(ctx, resGroup, serverName, encryptionProtectorObject)
+	if err != nil {
+		return err
+	}
+
+	if err = futureEncryptionProtector.WaitForCompletionRef(ctx, encryptionProtectorClient.Client); err != nil {
+		return fmt.Errorf("waiting on create/update future for Encryption Protector on server %q (Resource Group %q): %+v", serverName, resGroup, err)
+	}
+
+	// Encryption protector always uses "current" for the name
+	id := parse.NewEncryptionProtectorID(serverId.SubscriptionId, serverId.ResourceGroup, serverId.Name, "current")
+	d.SetId(id.ID())
+
+	return resourceMsSqlTransparentDataEncryptionRead(d, meta)
+}
+
+func resourceMsSqlTransparentDataEncryptionRead(d *schema.ResourceData, meta interface{}) error {
+	encryptionProtectorClient := meta.(*clients.Client).MSSQL.EncryptionProtectorClient
+
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.EncryptionProtectorID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := encryptionProtectorClient.Get(ctx, id.ResourceGroup, id.ServerName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request for Encryption Protector on server %s: %v", id.ServerName, err)
+	}
+
+	serverId := parse.NewServerID(id.SubscriptionId, id.ResourceGroup, id.ServerName)
+	d.Set("server_id", serverId.ID())
+
+	log.Printf("[INFO] Encryption protector key type is %s", resp.EncryptionProtectorProperties.ServerKeyType)
+
+	// Only set the key type if it's an AKV key. For service managed, we can omit the setting the key_vault_key_id
+	if resp.EncryptionProtectorProperties.ServerKeyType == sql.AzureKeyVault {
+		log.Printf("[INFO] Setting Key Vault URI to %s", *resp.EncryptionProtectorProperties.URI)
+
+		if err := d.Set("key_vault_key_id", resp.EncryptionProtectorProperties.URI); err != nil {
+			return fmt.Errorf("setting key_vault_key_id`: %+v", err)
+		}
+	}
+
```

Comment:
I suspect we'll need to set this in all cases to correctly identify diffs, and we'll need to nil check the `EncryptionProtectorProperties` to make sure we don't get a panic if it's missing for any reason.
```suggestion

	keyVaultKeyId := ""
	if resp.EncryptionProtectorProperties != nil && resp.EncryptionProtectorProperties.ServerKeyType  == sql.AzureKeyVault {
		if resp.EncryptionProtectorProperties.URL != nil {
			log.Printf("[INFO] Setting Key Vault URI to %s", *resp.EncryptionProtectorProperties.URI)
			keyVaultKeyId = *resp.EncryptionProtectorProperties.URI
		}
	}
	
	if err := d.Set("key_vault_key_id", keyVaultKeyId); err != nil {
		return fmt.Errorf("setting key_vault_key_id`: %+v", err)
	}

```

Reply:
@jackofallops on this one, does this now mean that key_vault_key_id would be empty string (that can be the default value) when the user wants to specify Service Managed and it would have a valid key vault key id when they want to use AKV? If so, I'm guessing that also means I'd have to remove the validationFunc that ensures key_vault_key_id is a valid KeyVault NestedItemId?